From 323b8cd6842d0557aa6ef35dca3e5b763b55c6e3 Mon Sep 17 00:00:00 2001
From: John Greene <jogreene@redhat.com>
Date: Wed, 10 Sep 2014 18:28:40 -0400
Subject: [ethernet] ixgbe: Fix up some ethtool results when adapter is removed

Message-id: <1410373750-11341-21-git-send-email-jogreene@redhat.com>
Patchwork-id: 93528
O-Subject: [RHEL7.1 PATCH 20/50] ixgbe: Fix up some ethtool results when adapter is removed
Bugzilla: 1091123
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Neil Horman <nhorman@redhat.com>

BZ: 1091123
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=7919168
Tested: In beaker. See BZ for details

Some ethtool tests returned apparently good results when the
adapter was in a removed state. Fix that by checking for removal.
This also fixes two paths that could return uninitialized memory
in data[4].

Signed-off-by: Mark Rustad <mark.d.rustad@intel.com>
Tested-by: Phil Schmitt <phillip.j.schmitt@intel.com>
Signed-off-by: Aaron Brown <aaron.f.brown@intel.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
(cherry picked from commit 0edd2bdf63b10d0b6715f2fc4a9e598d6b8ca58f)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/intel/ixgbe/ixgbe_ethtool.c b/drivers/net/ethernet/intel/ixgbe/ixgbe_ethtool.c
index 1222dc2..063166b 100644
--- a/drivers/net/ethernet/intel/ixgbe/ixgbe_ethtool.c
+++ b/drivers/net/ethernet/intel/ixgbe/ixgbe_ethtool.c
@@ -1,7 +1,7 @@
 /*******************************************************************************
 
   Intel 10 Gigabit PCI Express Linux driver
-  Copyright(c) 1999 - 2013 Intel Corporation.
+  Copyright(c) 1999 - 2014 Intel Corporation.
 
   This program is free software; you can redistribute it and/or modify it
   under the terms and conditions of the GNU General Public License,
@@ -1182,6 +1182,11 @@ static int ixgbe_link_test(struct ixgbe_adapter *adapter, u64 *data)
  struct ixgbe_hw *hw = &adapter->hw;
  bool link_up;
  u32 link_speed = 0;
+
+ if (ixgbe_removed(hw->hw_addr)) {
+  *data = 1;
+  return 1;
+ }
  *data = 0;
 
  hw->mac.ops.check_link(hw, &link_speed, &link_up, true);
@@ -1904,6 +1909,7 @@ static void ixgbe_diag_test(struct net_device *netdev,
   data[1] = 1;
   data[2] = 1;
   data[3] = 1;
+  data[4] = 1;
   eth_test->flags |= ETH_TEST_FL_FAILED;
   return;
  }
@@ -1923,6 +1929,7 @@ static void ixgbe_diag_test(struct net_device *netdev,
      data[1] = 1;
      data[2] = 1;
      data[3] = 1;
+     data[4] = 1;
      eth_test->flags |= ETH_TEST_FL_FAILED;
      clear_bit(__IXGBE_TESTING,
         &adapter->state);
-- 
1.7.1